kratos/elements/levelset_convection_element_simplex_algebraic_stabilization.h

```diff
@@ -0,0 +1,434 @@
+        const Variable<double>& rUnknownVar = my_settings->GetUnknownVariable();
+        const Variable<array_1d<double, 3 > >& rConvVar = my_settings->GetConvectionVariable();
+        const Variable<array_1d<double, 3 > >& rGradVar = my_settings->GetGradientVariable();
+
+        //getting data for the given geometry
+        BoundedMatrix<double, TNumNodes, TDim > DN_DX;
+        array_1d<double, TNumNodes > N;
+        double Volume;
+        GeometryUtils::CalculateGeometryData(GetGeometry(), DN_DX, N, Volume);
+
+        //here we get all the variables we will need
+        array_1d<double,TNumNodes> phi, phi_old;
+        array_1d< array_1d<double,3 >, TNumNodes> v, vold;
+
+        array_1d<double,3 > X_mean_tmp = ZeroVector(3);
+        array_1d<double,3 > grad_phi_mean_tmp = ZeroVector(3);
+        array_1d< array_1d<double,3 >, TNumNodes> X_node;
+        double phi_mean_old = 0.0;
+        double phi_mean = 0.0;
+
+        for (unsigned int i = 0; i < TNumNodes; i++)
+        {
+            phi[i] = GetGeometry()[i].FastGetSolutionStepValue(rUnknownVar);
+            phi_old[i] = GetGeometry()[i].FastGetSolutionStepValue(rUnknownVar,1);
+
+            v[i] = GetGeometry()[i].FastGetSolutionStepValue(rConvVar);
+            vold[i] = GetGeometry()[i].FastGetSolutionStepValue(rConvVar,1);
+
+            X_mean_tmp += GetGeometry()[i].Coordinates();
+            X_node[i] = GetGeometry()[i].Coordinates();
+
+            grad_phi_mean_tmp += GetGeometry()[i].GetValue(rGradVar);
+
+            phi_mean_old += GetGeometry()[i].FastGetSolutionStepValue(rUnknownVar,1);
+            phi_mean += GetGeometry()[i].FastGetSolutionStepValue(rUnknownVar);
+        }
+
+        phi_mean /= static_cast<double>(TNumNodes);
+        phi_mean_old /= static_cast<double>(TNumNodes);
+
+        array_1d<double,TDim> X_mean, grad_phi_mean;
+        for(unsigned int k = 0; k < TDim; k++)
+        {
+            grad_phi_mean[k] = grad_phi_mean_tmp[k]/static_cast<double>(TNumNodes);
+            X_mean[k] = X_mean_tmp[k]/static_cast<double>(TNumNodes);
+        }
+
+        array_1d<double,TDim> grad_phi_diff = prod(trans(DN_DX), phi_old) - grad_phi_mean;
+
+        BoundedMatrix<double,TNumNodes, TNumNodes> K_matrix = ZeroMatrix(TNumNodes, TNumNodes); // convection
+        BoundedMatrix<double,TNumNodes, TNumNodes> S_matrix = ZeroMatrix(TNumNodes, TNumNodes); // LHS stabilization
+        Vector S_vector = ZeroVector(TNumNodes); // RHS stabilization
+        BoundedMatrix<double,TNumNodes, TNumNodes> Mc_matrix = ZeroMatrix(TNumNodes, TNumNodes); // consistent mass matrix
+        BoundedMatrix<double,TNumNodes, TNumNodes> Ml_matrix = ZeroMatrix(TNumNodes, TNumNodes); // lumped mass matrix
+
+        BoundedMatrix<double,TNumNodes, TNumNodes> Ncontainer;
+        GetShapeFunctionsOnGauss(Ncontainer);
+        for(unsigned int igauss=0; igauss<TDim+1; igauss++)
```

Comment:
```suggestion
        for(unsigned int igauss=0; igauss<TDim+1; ++igauss)
```
This is (in theory) slightly more efficient and also assumed to be a good practice. In any case I don't think it makes any difference in this case, aside of the fact that most probably the compiler takes care.

Reply:
in case of [POD](https://en.cppreference.com/w/cpp/types/is_pod) types it shouldnt make any difference

I agree thought that it is good practice as it does make a difference when using iterators (there postincrement would create a copy)